GHC 8.8.4 fails sanity checks during install due to Cabal-3.0.1.0 being hidden

### General summary/comments

The LTS resolvers for GHC 8.8.4 (currently 16.12–16.17) fail to install GHC 8.8.4 in my environment because GHC complains about Cabal 3.0.1.0 being hidden when trying to build the sanity checks. This is the output I get when running `stack setup` in a freshly-created Stack project:

```
$ stack setup
Preparing to install GHC (tinfo6) to an isolated location.
This will not interfere with any system-level installation.
Downloaded ghc-tinfo6-8.8.4.
Installed GHC.
The GHC located at /home/zoe/.stack/programs/x86_64-linux/ghc-tinfo6-8.8.4/bin/ghc-8.8.4 failed to compile a sanity check. Please see:

    http://docs.haskellstack.org/en/stable/install_and_upgrade/

for more information. Exception was:
Received ExitFailure 1 when running
Raw command: /home/zoe/.stack/programs/x86_64-linux/ghc-tinfo6-8.8.4/bin/ghc-8.8.4 /tmp/stack-sanity-check-28c750373cb78094/Main.hs -no-user-package-db
Run from: /tmp/stack-sanity-check-28c750373cb78094/
Standard output:

[1 of 1] Compiling Main             ( /tmp/stack-sanity-check-28c750373cb78094/Main.hs, /tmp/stack-sanity-check-28c750373cb78094/Main.o )
Standard error:

Loaded package environment from /home/zoe/.ghc/x86_64-linux-8.8.4/environments/default

/tmp/stack-sanity-check-28c750373cb78094/Main.hs:1:1: error:
    Could not load module âDistribution.Simpleâ
    It is a member of the hidden package âCabal-3.0.1.0â.
    You can run â:set -package Cabalâ to expose it.
    (Note: this unloads all the modules in the current scope.)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
1 | import Distribution.Simple
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

This is in an up-to-date Arch Linux environment. Neither the nightly resolver nor the latest LTS for 8.8.3 (16.11) turn up this issue when I run `stack setup`. `Cabal-3.0.1.0` is present in `~/.stack/programs/x86_64-linux/ghc-tinfo6-8.8.4/lib/ghc-8.8.4`.

### Steps to reproduce

1. Create a new Stack project via `stack new`.
2. `cd` into the project directory.
3. Run `stack setup`.

### Expected

I expected to see `Installed GHC` followed by the `stack will use a sandboxed GHC it installed` etc. message.

### Actual

See [here](https://gist.github.com/spinnylights/77d0a1277df5a0a6beef9dab3fe8a820) for output from `stack setup --verbose`.

### Stack version

```
$ stack --version
Version 2.3.3, Git revision cb44d51bed48b723a5deb08c3348c0b3ccfc437e x86_64 hpack-0.33.0
```

### Method of installation

- [x] Official binary, downloaded from `get.haskellstack.org` to `~/.local/bin`

--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -22,6 +22,9 @@ Bug fixes:
 
 * Restore building of Stack with Cabal flag `disable-git-info` (broken with
   Stack 2.11.1).
+* Stack's sanity check on a selected GHC now passes GHC flag
+  `-hide-all-packages`, stopping GHC from looking for a package environment in
+  default locations.
 
 ## v2.11.1 - 2023-05-18
 

--- a/src/Stack/Setup.hs
+++ b/src/Stack/Setup.hs
@@ -2336,6 +2336,12 @@ sanityCheck ghc = withSystemTempDir "stack-sanity-check" $ \dir -> do
   eres <- withWorkingDir (toFilePath dir) $ proc (toFilePath ghc)
     [ fp
     , "-no-user-package-db"
+    -- Required to stop GHC looking for a package environment in default
+    -- locations.
+    , "-hide-all-packages"
+    -- Required because GHC flag -hide-all-packages is passed.
+    , "-package base"
+    , "-package Cabal" -- required for "import Distribution.Simple"
     ] $ try . readProcess_
   case eres of
     Left e -> prettyThrowIO $ GHCSanityCheckCompileFailed e ghc

